complex/cargo/list/src/lib.rs

```diff
@@ -37,228 +49,778 @@ pub struct ListManager {
-        let db = self.store.write_connection();
-        db.execute(sql, &[&name, &color]).unwrap();
-        self.fetch_label(&name)
+    pub fn create_labels_table(&mut self) -> Result<(), list_errors::Error> {
+        let schema = r#"[{  :db/ident     :label/name
+    :db/valueType :db.type/string
+    :db/cardinality :db.cardinality/one
+    :db/unique :db.unique/identity
+    :db/fulltext true },
+ {  :db/ident     :label/color
+    :db/valueType :db.type/string
+    :db/cardinality :db.cardinality/one }]"#;
+        let _ = self.write_connection().transact(schema)?;
+        Ok(())
     }
 
-    pub fn fetch_label(&self, name: &String) -> Option<Label> {
-        let sql = r#"SELECT name, color FROM labels WHERE name=?"#;
+    pub fn create_label(&mut self, name: String, color: String) -> Result<Option<Label>, list_errors::Error> {
+        let query = format!("[{{ :label/name \"{0}\" :label/color \"{1}\" }}]", &name, &color);
+        self.write_connection().transact(&query)?;
+        self.fetch_label(&name)
+    }
 
-        let conn = self.store.read_connection();
-        let mut stmt = conn.prepare(sql).unwrap();
-        let mut label_iter = stmt.query_map(&[name], |row| {
-            Label {
-                name: row.get(0),
-                color: row.get(1)
+    pub fn fetch_label(&self, name: &String) -> Result<Option<Label>, list_errors::Error> {
+        let query = r#"[:find ?eid, ?name, ?color
+            :in ?name
+            :where
+            [?eid :label/name ?name]
+            [?eid :label/color ?color]
+        ]"#;
+        let result = Arc::clone(&self.store).query_args(query, &[&(&"?name".to_string(), &name)])?;
+        if let QueryResults::Rel(rows) = result {
+            if !rows.is_empty() {
+                Ok(Label::from_row(&rows[0]))
+            } else {
+                Ok(None)
             }
-        }).unwrap();
-
-        if let Some(result) = label_iter.next() {
-            result.ok()
         } else {
-            println!("No label found for name {:?}", name);
-            None
+            Ok(None)
         }
     }
 
-    pub fn fetch_labels(&self) -> Vec<Label> {
-        let sql = r#"SELECT name, color
-                     FROM labels"#;
-        let conn = self.store.read_connection();
-        let mut stmt = conn.prepare(sql).unwrap();
-        let mut label_iter = stmt.query_map(&[], |row| {
-            Label {
-                name: row.get(0),
-                color: row.get(1)
-            }
-        }).unwrap();
+    pub fn fetch_labels(&self) -> Result<Vec<Label>, list_errors::Error> {
+        let query = r#"[:find ?eid, ?name, ?color
+            :where
+            [?eid :label/name ?name]
+            [?eid :label/color ?color]
+        ]"#;
+        let result = Arc::clone(&self.store).query(query)?;
+        if let QueryResults::Rel(rows) = result {
+            Ok(rows.iter().map(|row| Label::from_row(&row).unwrap()).collect())
+        } else {
+            Ok(vec![])
```

Comment:
Your approach here means that screwing up the query might make the function silently return an empty vector.

Try it: put `[]` around the variables in the query to make it a tuple query!

Prefer chaining, even if you need to define a trait to 'cast' the QR:

```
return self.store.query(query)
             .map(to_rel_rows)
             .map(…)
```